diff -Naur mc-4.6.1/src/dialog.c mc-4.6.1_UBSI/src/dialog.c
--- mc-4.6.1/src/dialog.c	2005-05-27 10:19:18.000000000 -0400
+++ mc-4.6.1_UBSI/src/dialog.c	2016-05-20 13:21:07.062897196 -0400
@@ -15,7 +15,7 @@
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
-
+#include "UBSI.h"
 #include <config.h>
 #include <string.h>
 #include <stdio.h>
@@ -753,6 +753,7 @@
 
     event.x = -1;
     while (h->running) {
+						UBSI_LOOP_ENTRY(1);
 	if (winch_flag)
 	    change_screen_size ();
 
@@ -779,6 +780,7 @@
 	if (!h->running)
 	    (*h->callback) (h, DLG_VALIDATE, 0);
     }
+				UBSI_LOOP_EXIT(1);
 }
 
 /* Standard run dialog routine
diff -Naur mc-4.6.1/src/UBSI.h mc-4.6.1_UBSI/src/UBSI.h
--- mc-4.6.1/src/UBSI.h	1969-12-31 19:00:00.000000000 -0500
+++ mc-4.6.1_UBSI/src/UBSI.h	2016-05-20 13:17:59.610897196 -0400
@@ -0,0 +1,28 @@
+#ifndef UBSI_DEFINE
+#define UBSI_DEFINE
+
+#define UBSI_MEM_WRITE(addr) { \
+  intptr_t value = (intptr_t)addr; \
+  uint32_t high = value >> 32; \
+  uint32_t low = value; \
+  kill(-300, high); \
+  kill(-301, low); \
+}
+
+#define UBSI_MEM_READ(addr) { \
+  intptr_t value = (intptr_t)addr; \
+  uint32_t high = value >> 32; \
+  uint32_t low = value; \
+  kill(-200, high); \
+  kill(-201, low); \
+}
+
+#define UBSI_LOOP_ENTRY(loopId) { \
+  kill(-100, loopId); \
+}
+
+#define UBSI_LOOP_EXIT(loopId) { \
+  kill(-101, loopId); \
+}
+
+#endif
